### PR TITLE
Parser rewrite

### PR DESCRIPTION
**Pull request by lorenzo**
Took the time to do an almost-rewrite of the parser, with the benefits:

* No need to normalize lines. This means that errors are reported with a lot more precision
* Fixed parsing of env vars, especially around variables
* Added experimental `RUN` flags support
* Improved error messages for certain edge cases

Additionally:
* Moved the EDSL stuff out. This makes the package leaner and will help in cross-compiling hadolint to other architectures
* Added a few more tests

### CODE REVIEW

File: language-docker.cabal [hunk 1]
```diff
@@ -0,0 +1,90 @@
+cabal-version: 1.12
+
+-- This file has been generated from package.yaml by hpack version 0.31.2.
```
Comment (zemanlx) [inline language-docker.cabal:3]: Do you want to include this file when it is autogenerated?

File: src/Language/Docker/Parser/Run.hs [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+  reserved "RUN"
+  Run <$> runArguments
+
+runArguments :: Parser (RunArgs Text)
+runArguments = do
+  presentFlags <- choice [runFlags <* requiredWhitespace, pure (RunFlags Nothing Nothing Nothing)]
+  args <- arguments
+  return $ RunArgs args presentFlags
+
+runFlags :: Parser RunFlags
```
Comment (zemanlx) [inline src/Language/Docker/Parser/Run.hs:55]: Does this mean that we will support experimental `buildkit` syntax?  I have found it only here https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/experimental.md. Do you have some better resource?

File: language-docker.cabal (hunk 1 above)
Comment (lorenzo) [reply language-docker.cabal:3]: Yeah, the new recommendations from stack is to include it. It seems to avoid errors in versioning and makes life of people not using stack easier

File: src/Language/Docker/Parser/Run.hs (hunk 2 above)
Comment (lorenzo) [reply src/Language/Docker/Parser/Run.hs:55]: Yes we do support them now!

I used that file, but as usual with the dockerfile syntax, I ended up reading the source code to discover lots more.

Comment (zemanlx) [reply src/Language/Docker/Parser/Run.hs:55]: cool, thanks
